### PR TITLE
Updates ml.net reference of LightGBM to version 2.2

### LINKED ISSUE #2446: Updating LightGBM to version 2.2.3
LightGBM has been updated to version 2.2.3, ml.net references 2.2.1.1. 

Creating this issue to update ml.net to the latest version of LightGBM.

### PR DESCRIPTION
**Pull request by singlis**
- Updates ml.net reference of LightGBM to version 2.2.3 (Fixes #2446)
- Updated the lightgbm parsing code to handle inf, -inf (now checks for
contains rather than equals).
- Additional updates for handling NaN
- Moved all LightGBM baseline tests from SingleDebug/SingleRelease to
Common.
- Added Seed parameter to LightGBM arguments to support setting
LightGBM's random seed.

### CODE REVIEW

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs [hunk 1]
```diff
@@ -565,15 +571,24 @@ private void ExecuteTFTransformMNISTConvTrainingTest(bool shuffle, int? shuffleS
                     }))
                     .Append(mlContext.Transforms.Concatenate("Features", "Prediction"))
                     .AppendCacheCheckpoint(mlContext)
-                    .Append(mlContext.MulticlassClassification.Trainers.LightGbm("Label", "Features"));
+                    .Append(mlContext.MulticlassClassification.Trainers.LightGbm(new LightGBM.Options() {
+                        LabelColumn="Label", 
+                        FeatureColumn="Features", 
+                        Seed=1, 
+                        NThread=1, 
+                        NumBoostRound=1}));
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs]: ctrl + K +D aka formatting

File: src/Microsoft.ML.LightGBM/WrappedLightGbmBooster.cs [hunk 2]
```diff
@@ -100,12 +100,12 @@ private static double[] Str2DoubleArray(string str, char delimiter)
             var values = new List<double>();
             foreach (var token in str.Split(delimiter))
             {
-                var trimmed = token.Trim();
+                var trimmed = token.Trim().ToLowerInvariant();
 
-                if (trimmed.Equals("inf", StringComparison.OrdinalIgnoreCase))
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.LightGBM/WrappedLightGbmBooster.cs]: >StringComparison.OrdinalIgnoreCase [](start = 42, length = 34)

PUT IT BACK!!!111

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs [hunk 3]
```diff
@@ -565,15 +571,24 @@ private void ExecuteTFTransformMNISTConvTrainingTest(bool shuffle, int? shuffleS
...
 
                 var trainedModel = pipe.Fit(preprocessedTrainData);
                 var predicted = trainedModel.Transform(preprocessedTestData);
                 var metrics = mlContext.MulticlassClassification.Evaluate(predicted);
 
                 // First group of checks. They check if the overall prediction quality is ok using a test set.
-                Assert.InRange(metrics.AccuracyMicro, expectedMicroAccuracy - .01, expectedMicroAccuracy + .01);
-                Assert.InRange(metrics.AccuracyMacro, expectedMacroAccruacy - .01, expectedMicroAccuracy + .01);
+                bool inRange = expectedValues.Any(exp =>
```
Comment (wschin) [inline test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs]: Why not use `Assert.Equal` with a proper tolerance (e.g., 2) or `Assert.InRange`? They automatically print out error information.

Comment (singlis) [reply test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs]: I wanted to use Assert.InRange, however this is given a list of expected ranges that should match (i.e. micro and macro should match one of those ranges). Because of this, Assert.InRange will not work because it will fail on the first range that doesnt match. 

---
In reply to: [254510687](https://github.com/dotnet/machinelearning/pull/2448#discussion_r254510687) [](ancestors = 254510687)

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs [hunk 4]
```diff
@@ -500,11 +500,17 @@ private void CleanUp(string model_location)
         [ConditionalFact(typeof(Environment), nameof(Environment.Is64BitProcess))] // TensorFlow is 64-bit only
         public void TensorFlowTransformMNISTConvTrainingTest()
         {
-            ExecuteTFTransformMNISTConvTrainingTest(false, null, 0.74782608695652175, 0.608843537414966);
-            ExecuteTFTransformMNISTConvTrainingTest(true, 5, 0.75652173913043474, 0.610204081632653);
+            var expectedValues = new List<(double micro, double macro)>()
+            {
+                (0.71304347826086956, 0.53197278911564627),
+                (0.73304347826086956, 0.677551020408163)
```
Comment (zeahmed) [inline test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs]: so debug and release versions still differ?

**Comment by singlis** [comment]
Here are the differences between master and this PR:

> 
Flight Data
lightgbm update
OVERALL RESULTS
---------------------------------------
L1(avg):            0.299062 (0.0000)
L2(avg):            0.144591 (0.0000)
RMS(avg):           0.380251 (0.0000)
Loss-fn(avg):       0.144591 (0.0000)
R Squared:          0.069896 (0.0000)

---------------------------------------
Physical memory usage(MB): 113
Virtual memory usage(MB): 2101875
2/12/2019 1:11:53 AM     Time elapsed(s): 7.502

Master
OVERALL RESULTS
---------------------------------------
L1(avg):            0.299102 (0.0000)
L2(avg):            0.144667 (0.0000)
RMS(avg):           0.380351 (0.0000)
Loss-fn(avg):       0.144667 (0.0000)
R Squared:          0.069406 (0.0000)

---------------------------------------
Physical memory usage(MB): 112
Virtual memory usage(MB): 2101875
2/12/2019 12:29:38 AM    Time elapsed(s): 7.429


MNIST
LigthGBM Update
OVERALL RESULTS
---------------------------------------
Accuracy(micro-avg): 0.976700 (0.0000)
Accuracy(macro-avg): 0.976538 (0.0000)
Log-loss:           0.078717 (0.0000)
Log-loss reduction: 96.578795 (0.0000)

---------------------------------------
Physical memory usage(MB): 207
Virtual memory usage(MB): 2102375
2/12/2019 1:35:23 AM     Time elapsed(s): 63.784

Master
OVERALL RESULTS
---------------------------------------
Accuracy(micro-avg): 0.975400 (0.0000)
Accuracy(macro-avg): 0.975237 (0.0000)
Log-loss:           0.078511 (0.0000)
Log-loss reduction: 96.587753 (0.0000)

---------------------------------------
Physical memory usage(MB): 199
Virtual memory usage(MB): 2102378
2/12/2019 1:09:02 AM     Time elapsed(s): 63.62

**Comment by singlis** [comment]
Further benchmarks:
[Internal ranking dataset with 500k rows]
LightGBM Update
OVERALL RESULTS
---------------------------------------
AUC:                0.793242 (0.0000)
Accuracy:           0.881250 (0.0000)
Positive precision: 0.748541 (0.0000)
Positive recall:    0.273974 (0.0000)
Negative precision: 0.888696 (0.0000)
Negative recall:    0.984371 (0.0000)
Log-loss:           0.467847 (0.0000)
Log-loss reduction: 21.711629 (0.0000)
F1 Score:           0.401130 (0.0000)
AUPRC:              0.515647 (0.0000)

---------------------------------------
Physical memory usage(MB): 2472
Virtual memory usage(MB): 2104804
2/12/2019 1:32:10 AM     Time elapsed(s): 252


Master
OVERALL RESULTS
---------------------------------------
AUC:                0.793242 (0.0000)
Accuracy:           0.881250 (0.0000)
Positive precision: 0.748541 (0.0000)
Positive recall:    0.273974 (0.0000)
Negative precision: 0.888696 (0.0000)
Negative recall:    0.984371 (0.0000)
Log-loss:           0.467847 (0.0000)
Log-loss reduction: 21.711629 (0.0000)
F1 Score:           0.401130 (0.0000)
AUPRC:              0.515647 (0.0000)

---------------------------------------
Physical memory usage(MB): 2471
Virtual memory usage(MB): 2104846
2/12/2019 1:03:58 AM     Time elapsed(s): 224

Wiki Detox
LightGBM Update
iter = 100
OVERALL RESULTS
---------------------------------------
Accuracy(micro-avg): 0.958941 (0.0011)
Accuracy(macro-avg): 0.823435 (0.0049)
Log-loss:           0.114225 (0.0028)
Log-loss reduction: 63.930055 (0.5704)

---------------------------------------
Physical memory usage(MB): 8486
Virtual memory usage(MB): 2111485
2/11/2019 6:43:24 PM     Time elapsed(s): 956

Master
OVERALL RESULTS
---------------------------------------
Accuracy(micro-avg): 0.958194 (0.0009)
Accuracy(macro-avg): 0.817749 (0.0041)
Log-loss:           0.116529 (0.0028)
Log-loss reduction: 63.202381 (0.5569)

---------------------------------------
Physical memory usage(MB): 8561
Virtual memory usage(MB): 2111419
2/11/2019 11:50:51 PM    Time elapsed(s): 849


NOTE These are informal benchmarks to spotcheck if there are any regressions with the LightGBM Update. There are time differences on WIkiDetox and the internal ranking dataset as the LightGBM update shows a slightly longer time. This is likely due to running a single test on my work machine, but not concerning as the other benchmarks do fall more inline.
